### ISSUE TITLE
Training stops abruptly after every 4 epochs with streaming zipformer2 recipe

### ISSUE THREAD

**Issue by jeenajp** [k2-fsa/icefall#1454]
```
    train_dl_iter=iter(train_dl)
    #extra_dloaders = [iter(dloader_libri),iter(dload_gigaspeech),iter(dload_commonvoice)]

    batch_idx = 0
    previous_batch="open"

    while True:

        try:

            if previous_batch=="open":
                batch = next(train_dl_iter)
                previous_batch="main"
            else:
                chosen_dloader=random.choice(extra_train_dl)
                batch=next(chosen_dloader)
                previous_batch="open"


        except StopIteration:
            break

        if batch_idx % 10 == 0:
            set_batch_count(model, get_adjusted_batch_count(params))

        batch_idx += 1
        params.batch_idx_train += 1
        batch_size = len(batch["supervisions"]["text"])

        logging.info("compute_loss called by train_one_epoch- - ")


```
**Context**

We are using streaming zipformer2 recipe. But we made a change to handle multiple datasets during training - instead of using the same dataset, we are alternating between main dataset and one of the randomly chosen open source datasets every batch.
 
We face a problem with this setup. Training is stopping after every fourth epoch. For example, training ran fine till 4th epoch and stopped. Restarting from 5th epoch worked fine until 8th epoch after which it stopped. We tried computing WER with these models and the numbers look good. GPUs are getting maxed out. We did not notice any error message in logs.

**Comment by csukuangfj** [comment]
> We did not notice any error message in logs.

Are there any logs after the training stops?

Could you use `print` or `logging.warning` or `logging.error` to replace `logging.info` to get more logs if `logging.info` does not work?

**Comment by jeenajp** [comment]
There is nothing in log files specifically.
In one iteration where I started from 21st epoch, it ran fine till24th epoch but behaved weird from 25th epoch (did not stop but models were getting saved within 10 to 20 minutes). Log file corresponding to this is uploaded here. After last line in this, GPU got maxed out.
[log.txt](https://github.com/k2-fsa/icefall/files/13871785/log.txt)
 Logs are getting printed but we did not get any relevant information to decode this issue.
Attaching the screen shot to show the timestamps at which model files are created. We can see that the usual time taken for an epoch is around 1.5 hours but from 25th epoch the pattern doesn't make sense. 
![epochs_timestamps_image](https://github.com/k2-fsa/icefall/assets/123490082/632cd2b3-a8b6-4f1f-a69a-30858041bed3)

**Comment by marcoyang1998** [comment]
Are you using multiple dataloaders?

**Comment by jeenajp** [comment]
Yes, we are using multiple dataloaders. We will look into this and update you. Thanks.

**Comment by uni-saurabh-vyas** [comment]
Just for the sake of completion and benefit of others, I would like to point out 2 things that caused the issue.

1. We were using AWS SageMaker, and when this happened, the model was no longer getting trained, but, the training job kept continuing (this was caused because the raise statement was missed in the Exception block), which caused one of the process to not return the status code correctly.

2. Unlike main dataset loader, for the extra datasets, we were not creating an iter of the dataloader inside train_one_epoch function(we had created in some other function), as a result, the extra data loaders weren't getting reset. 

So basically not icefall related ...
